### DISCUSSION TITLE
When is an object "added" to ruby (and can be GCed) ?

### DISCUSSION THREAD

**Discussion by sasq64** [mruby/mruby#5468] [Q&A] [answered]
At which points in a program does Ruby take control over mrb_values created in C ?

If I just create an object, and don't return it to ruby somehow, it wont be freed by Ruby, so I need to free it myself.
But if the mrb_value is transfered to ruby somehow, I must make sure to _not_ free it myself.

I find it tricky to create a strategy around objects created in native.

**Comment by matz** [comment, marked as answer]
The objects created by `mrb_obj_alloc()` directly or indirectly are all managed by GC. When they are allocated they are also referenced from `GC arena`, which keeps object reference until C function terminates. See [doc/guides/gc-arena-howto.md](https://github.com/mruby/mruby/blob/master/doc/guides/gc-arena-howto.md)
